### PR TITLE
Introducing embed(url) root query and api auth and access for Vimeo, …

### PR DESCRIPTION
**Pull request by lukasbuenger**
…Youtube and Twitter.

##### Changelog

* Extends the schema with
   * `interface EmbedInterface`
   * `type Twitter implements EmbedInterface`
   * `type Vimeo implements EmbedInterface`
   * `type Youtube implements EmbedInterface`
   * `union Embed = Twitter | Vimeo | Youtube`
   * Root query `embed(url: String!): Embed!`
* Adds resolvers for the `embed` query and the `Embed` union type
* Introduces `lib/twitter`, `lib/vimeo` and `lib/youtube`. They implement auth and retrieval for the respective entities, following the guidelines suggested by `lib/github`. 
* Expects the following .env values:
   * `TWITTER_APP_KEY`
   * `TWITTER_APP_SECRET`
   * `VIMEO_APP_KEY`
   * `VIMEO_APP_SECRET`
   * `YOUTUBE_APP_KEY`

### CODE REVIEW

**Comment by lukasbuenger** [comment]
FYI: The build is failing and raises the same errors I had, when trying to add tests. @patte Any ideas what I'm missing here? I'd gladly add the integration test coverage once I've figured out what's going wrong here.

File: graphql/resolvers/Embed.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+module.exports = {
+  __resolveType (obj) {
+    // obj is the entity from the DB and thus has the "rewardType" column used as FK
```
Comment (patte) [inline graphql/resolvers/Embed.js]: remove this line

File: graphql/resolvers/RootQuerys/embed.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
+const { ensureUserHasRole } = require('../../../lib/Roles')
+const { getTweetById } = require('../../../lib/twitter')
+const { getYoutubeVideoById } = require('../../../lib/youtube')
+const { getVimeoVideoById } = require('../../../lib/vimeo')
+
+const TWITTER_REGEX = /^https?:\/\/twitter\.com\/(?:#!\/)?(\w+)\/status(es)?\/(\d+)$/
+const YOUTUBE_REGEX = /^.*(?:(?:youtu\.be\/|v\/|vi\/|u\/\w\/|embed\/)|(?:(?:watch)?\?v(?:i)?=|&v(?:i)?=))([^#&?]*).*/
+const VIMEO_REGEX = /(http|https)?:\/\/(www\.)?vimeo.com\/(?:channels\/(?:\w+\/)?|groups\/([^/]*)\/videos\/|)(\d+)(?:|\/\?)/
```
Comment (patte) [inline graphql/resolvers/RootQuerys/embed.js]: maybe note what group wraps what...

File: graphql/schema.js [hunk 3]
```diff
@@ -206,5 +207,43 @@ input DocumentInput {
...
+  id: ID!
+  url: String!
+
+  createdAt: String!
+  userId: String!
+  userName: String!
+  thumbnail: String!
+}
+
+union Embed = Twitter | Youtube | Vimeo
```
Comment (patte) [inline graphql/schema.js]: IMHO `union Embed = TwitterEmbed | YoutubeEmbed | VimeoEmbed` would be better language.

File: lib/twitter/appAuth.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
+/**
```
Comment (patte) [inline lib/twitter/appAuth.js]: rename this to auth.js or oauth.js

File: lib/vimeo/appAuth.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
+/**
```
Comment (patte) [inline lib/vimeo/appAuth.js]: rename this to auth.js or oauth.js

**Comment by patte** [comment]
It would be cool to mention at least the URL in the README to where to get the API keys for youtube, vimeo and twitter.

**Comment by patte** [comment]
pull from master to make tests pass again.

**Comment by lukasbuenger** [comment]
Ok, the last structural refactoring:
Now, the embed endpoint takes an id and an embedType and returns accordingly.

I added a couple of tests that pass on my machine, but somehow the process falls asleep and doesn't recover. Let's see what Travis has to say ...

**Comment by patte** [comment]
one last question: you don't check if the tokens are expired. how long are they valid?

**Comment by lukasbuenger** [comment]
Neither of the 3 tokens I introduced with this pull request work based on expiry. This is a quote from the Twitter docs and it does quite represent the current thinking about access keys and tokens in social media. This is not Sparta.

>Access tokens are not explicitly expired. An access token will be invalidated if a user explicitly revokes an application in the their Twitter account settings, or if Twitter suspends an application. If an application is suspended, there will be a note on the apps.twitter.com page stating that it has been suspended. 

[Source](https://developer.twitter.com/en/docs/basics/authentication/guides/oauth-faq)

**Comment by patte** [comment]
nice work, with tests! ilike

**Pull request merged**